![](../../images/featuretools.png)

# Featuretools Implementation with Dask

A calculation of Deep Feature Synthesis from the Automated Loan Repayment notebook running on a single core takes about 25 hours on an AWS EC2 machine. Clearly we need a better approach for practical implementations of calculating a large feature matrix, one that allows us to use all cores of whatever machine we are using. 

Featuretools does have support for parallel processing if you have multiple cores (which nearly every laptop now does), but it currently sends the entire EntitySet to each process, which means you might exhaust the memory on any one core. For example, the AWS machine has 8 GB per core, which might seem like a lot until you realize the EntitySet takes up about 11 GB and setting `n_jobs=-1` will cause an out of memory error. Therefore, we cannot use the parallel processing in Featuretools and instead have to engineer our own implementation with Dask.  

Fortunately, options such as [Dask](https://dask.pydata.org/en/latest/) make it easy to take advantage of multiple cores on our own machine. In this notebook, we'll see how to run Deep Feature Synthesis in about 3 hours on a personal laptop with 8 cores and 16 GB of RAM. 

![](../../images/dask_logo.png)


## Roadmap

Following is our plan of action for implementing Dask with Featuretools

1. Convert `object` data types to `category`
    * This reduces memory consumption significantly
2. Create 104 partitions of data and save to disk
    * Each partition will contain all data for 1/104 of the clients
    * Each partition can be used to __independently__ make an EntitySet and then a feature matrix
3. Write a function to take a partition and create an `EntitySet`
4. Write a function to take an `EntitySet` and calculate a `feature_matrix` that is saved to disk
5. Use Dask to parallelize 3. and 4. to create 104 feature matrices saved on disk
6. (Optionally) read in the individual feature matrices and combine into a single feature matrix
    
The general idea is to __take advantage of all our system resources by breaking one large problem into many smaller ones.__ Each of these smaller problems can be completed on one processor which means we can run multiple (8) of these problems at a time. 

At the end, we'll have a working implementation of Dask that lets us take full advantage of our computing resources. While a naive approach to this problem would just be renting a larger machine, that won't solve our problem because the bottleneck is not RAM, but using multiple cores at a time. __Often, a better approach than getting more computating resources is to find a way to use the resources we do have as efficiently as possible.__ 

In [1]:
# pandas and numpy for data manipulation
import pandas as pd
import numpy as np

# featuretools for automated feature engineering
import featuretools as ft
import featuretools.variable_types as vtypes

# Utilities
import sys
import psutil
import os

from timeit import default_timer as timer

## Convert Data Types

The first step is to convert all the data types we can. Using `category` instead of `object` can significantly reduced memory usage if the number of unique categories is much less than the number of observations. For more on the `category` type in Pandas, look at [the documentation.](https://pandas.pydata.org/pandas-docs/stable/categorical.html)

While this isn't specific to Dask, it's a good practice in general. The function below can be modified for different problems as required.

In [2]:
def convert_types(df):
    """Convert pandas data types for memory reduction."""
    
    # Iterate through each column
    for c in df:
        
        # Convert ids and booleans to integers
        if ('SK_ID' in c):
            df[c] = df[c].fillna(0).astype(np.int32)
            
        # Convert objects to category
        elif (df[c].dtype == 'object') and (df[c].nunique() < df.shape[0]):
            df[c] = df[c].astype('category')
        
        # Booleans mapped to integers
        elif set(df[c].unique()) == {0, 1}:
            df[c] = df[c].astype(bool)
        
        # Float64 to float32
        elif df[c].dtype == float:
            df[c] = df[c].astype(np.float32)
            
        # Int64 to int32
        elif df[c].dtype == int:
            df[c] = df[c].astype(np.int32)
        
    return df

Now we'll read in the datasets and apply the convert types function. 

In [3]:
# Read in the datasets and replace the anomalous values
app_train = pd.read_csv('../input/application_train.csv').replace({365243: np.nan})
app_test = pd.read_csv('../input/application_test.csv').replace({365243: np.nan})
bureau = pd.read_csv('../input/bureau.csv').replace({365243: np.nan})
bureau_balance = pd.read_csv('../input/bureau_balance.csv').replace({365243: np.nan})
cash = pd.read_csv('../input/POS_CASH_balance.csv').replace({365243: np.nan})
credit = pd.read_csv('../input/credit_card_balance.csv').replace({365243: np.nan})
previous = pd.read_csv('../input/previous_application.csv').replace({365243: np.nan})
installments = pd.read_csv('../input/installments_payments.csv').replace({365243: np.nan})

app_test['TARGET'] = np.nan

# Join together training and testing
app = app_train.append(app_test, ignore_index = True, sort = True)
number_clients = app.shape[0]

# Need `SK_ID_CURR` in every dataset
bureau_balance = bureau_balance.merge(bureau[['SK_ID_CURR', 'SK_ID_BUREAU']], 
                                      on = 'SK_ID_BUREAU', how = 'left')

print(f"""Total memory before converting types: \
{round(np.sum([x.memory_usage().sum() / 1e9 for x in 
[app, bureau, bureau_balance, cash, credit, previous, installments]]), 2)} gb.""")

# Convert types to reduce memory usage
app = convert_types(app)
bureau = convert_types(bureau)
bureau_balance = convert_types(bureau_balance)
cash = convert_types(cash)
credit = convert_types(credit)
previous = convert_types(previous)
installments = convert_types(installments)

print(f"""Total memory after converting types: \
{round(np.sum([x.memory_usage().sum() / 1e9 for x in 
[app, bureau, bureau_balance, cash, credit, previous, installments]]), 2)} gb.""")

# Set the index for locating
for dataset in [app, bureau, bureau_balance, cash, credit, previous, installments]:
    dataset.set_index('SK_ID_CURR', inplace = True)

Total memory before converting types: 4.38 gb.
Total memory after converting types: 2.03 gb.


In [4]:
print('Object memory usage.')
print(bureau['CREDIT_TYPE'].astype('object').memory_usage() / 1e9, 'gb')

print('Category memory usage.')
print(bureau['CREDIT_TYPE'].astype('category').memory_usage() / 1e9, 'gb')

print('Length of data: ', bureau.shape[0])
print('Number of unique categories: ', bureau['CREDIT_TYPE'].nunique())

Object memory usage.
0.027462848 gb
Category memory usage.
0.015448612 gb
Length of data:  1716428
Number of unique categories:  15


We can see the significant difference in memory usage depending on the data type. Since we are looking to get the most from our machine, any step that can reduce computational overhead is beneficial.

# Partitioning Data

Next, we partition the data into 104 separate datasets based on the client id, `SK_ID_CURR` and save the partitions to disk. Every partition will contain the data associated with a subset of the clients and therefore will have 7 smaller csv files. __What is important here is that each partition contains all the data needed to make a feature matrix for the clients and therefore the feature matrix calculations are independent of one another.__

* Each partition by itself contains all the data needed to make an `EntitySet` for the clients 
* This `EntitySet` can then be used to create a feature matrix 
* Partitioning and saving the raw data allows for more flexilibilitiy when we create the entity set and feature matrix

104 partitions was chosen after some trial and error guided by 4 general ideas:

1. We want more tasks than workers
2. The tasks must be small enough that they don't exhaust the memory of an individual worker
3. More tasks will decrease the variance in completion time for each task
4. The number of tasks should be a multiple of the number of workers

As with many applications in machine learning, experimentation is often the best way to find what works.

In [5]:
def create_partition(user_list, partition):
    """Creates and saves a dataset with only the users in `user_list`."""
    
    # Make the directory
    directory = '../input/partitions/p%d' % (partition + 1)
    if os.path.exists(directory):
        return
    
    else:
        os.makedirs(directory)
        
        # Subset based on user list
        app_subset = app[app.index.isin(user_list)].copy().reset_index()
        bureau_subset = bureau[bureau.index.isin(user_list)].copy().reset_index()

        # Drop SK_ID_CURR from bureau_balance, cash, credit, and installments
        bureau_balance_subset = bureau_balance[bureau_balance.index.isin(user_list)].copy().reset_index(drop = True)
        cash_subset = cash[cash.index.isin(user_list)].copy().reset_index(drop = True)
        credit_subset = credit[credit.index.isin(user_list)].copy().reset_index(drop = True)
        previous_subset = previous[previous.index.isin(user_list)].copy().reset_index()
        installments_subset = installments[installments.index.isin(user_list)].copy().reset_index(drop = True)
        

        # Save data to the directory
        app_subset.to_csv('%s/app.csv' % directory, index = False)
        bureau_subset.to_csv('%s/bureau.csv' % directory, index = False)
        bureau_balance_subset.to_csv('%s/bureau_balance.csv' % directory, index = False)
        cash_subset.to_csv('%s/cash.csv' % directory, index = False)
        credit_subset.to_csv('%s/credit.csv' % directory, index = False)
        previous_subset.to_csv('%s/previous.csv' % directory, index = False)
        installments_subset.to_csv('%s/installments.csv' % directory, index = False)

        if partition % 10 == 0:
            print('Saved all files in partition {} to {}.'.format(partition + 1, directory))

In [6]:
# Break into 104 chunks
chunk_size = app.shape[0] // 103

# Construct an id list
id_list = [list(app.iloc[i:i+chunk_size].index) for i in range(0, app.shape[0], chunk_size)]

In [7]:
from itertools import chain

# Sanity check that we have not missed any ids
print('Number of ids in id_list:         {}.'.format(len(list(chain(*id_list)))))
print('Total length of application data: {}.'.format(len(app)))

Number of ids in id_list:         356255.
Total length of application data: 356255.


In [8]:
start = timer()
for i, ids in enumerate(id_list):
    # Create a partition based on the ids
    create_partition(ids, i)
    
end = timer()
print(f'Partitioning took {round(end - start)} seconds.')

Partitioning took 0 seconds.


__I already had the partitions made, but running the above cell took 1300 seconds (21 minutes) the first time. __

We can independently generate the feature matrix for each partition because the partition contains all the data for that group of clients. Moreover, each subset of data is small enough for the feature matrix calculation to fit entirely on one core.

#### Load in Feature Definitions

We already calculated the feature definitions in the Automated Loan Repayment notebook so we can read them in. This avoids the need to have to recalculate the features on each partition. Instead of using `ft.dfs`, if we have the feature names, we can use `ft.calculate_feature_matrix` and pass in the `EntitySet` and the feature names. More importantly, it ensures that we create __the exact same set of features for each partition.__

In [9]:
feature_defs = ft.load_features('../input/features.txt')
print(len(feature_defs))

1820


For each feature matrix, we'll make 1820 features.

#### Variable Types

If the Automated notebook, we specified the variable types when adding entities to the entityset. However, since we already properly defined the data types for each column, Featuretools will now infer the correct variable type. For example, while before we have Booleans mapped to integers which would be interpreted as numeric, now the Booleans are represented as Booleans and hence will be correctly inferred by Featuretools.

# Function to Create EntitySet from Partition 

The next function takes a single partition of data and make an `EntitySet`. We won't save these entitysets to disk, but instead will keep them in memory while calculating the feature matrices. Therefore, if we want to make any changes to the `EntitySet`, such as adding in interesting values or seed features, we can alter this function and remake the `EntitySet` without having to rewrite all the Entity Sets on disk. Writing the entity sets to disk would be another option if we are sure that they won't ever change. We are not going to change the raw data which is why I choose to save the data partitions to the hard drive.

In [10]:
def entityset_from_partition(path):
    """Create an EntitySet from a partition of data specified as a path.
       Returns a dictionary with the entityset and the number used for saving the feature matrix."""
    
    partition_num = int(path[21:])
    
    # Read in data
    app = pd.read_csv('%s/app.csv' % path)
    bureau = pd.read_csv('%s/bureau.csv' % path)
    bureau_balance = pd.read_csv('%s/bureau_balance.csv' % path)
    previous = pd.read_csv('%s/previous.csv' % path)
    credit = pd.read_csv('%s/credit.csv' % path)
    installments = pd.read_csv('%s/installments.csv' % path)
    cash = pd.read_csv('%s/cash.csv' % path)
    
    # Empty entityset
    es = ft.EntitySet(id = 'clients')
    
    # Entities with a unique index
    es = es.entity_from_dataframe(entity_id = 'app', dataframe = app, index = 'SK_ID_CURR')

    es = es.entity_from_dataframe(entity_id = 'bureau', dataframe = bureau, index = 'SK_ID_BUREAU')

    es = es.entity_from_dataframe(entity_id = 'previous', dataframe = previous, index = 'SK_ID_PREV')

    # Entities that do not have a unique index
    es = es.entity_from_dataframe(entity_id = 'bureau_balance', dataframe = bureau_balance, 
                                  make_index = True, index = 'bureaubalance_index')

    es = es.entity_from_dataframe(entity_id = 'cash', dataframe = cash, 
                                  make_index = True, index = 'cash_index')

    es = es.entity_from_dataframe(entity_id = 'installments', dataframe = installments,
                                  make_index = True, index = 'installments_index')

    es = es.entity_from_dataframe(entity_id = 'credit', dataframe = credit,
                                  make_index = True, index = 'credit_index')
    
    # Relationship between app_train and bureau
    r_app_bureau = ft.Relationship(es['app']['SK_ID_CURR'], es['bureau']['SK_ID_CURR'])

    # Relationship between bureau and bureau balance
    r_bureau_balance = ft.Relationship(es['bureau']['SK_ID_BUREAU'], es['bureau_balance']['SK_ID_BUREAU'])

    # Relationship between current app and previous apps
    r_app_previous = ft.Relationship(es['app']['SK_ID_CURR'], es['previous']['SK_ID_CURR'])

    # Relationships between previous apps and cash, installments, and credit
    r_previous_cash = ft.Relationship(es['previous']['SK_ID_PREV'], es['cash']['SK_ID_PREV'])
    r_previous_installments = ft.Relationship(es['previous']['SK_ID_PREV'], es['installments']['SK_ID_PREV'])
    r_previous_credit = ft.Relationship(es['previous']['SK_ID_PREV'], es['credit']['SK_ID_PREV'])
    
    # Add in the defined relationships
    es = es.add_relationships([r_app_bureau, r_bureau_balance, r_app_previous,
                               r_previous_cash, r_previous_installments, r_previous_credit])

    return ({'es': es, 'num': partition_num})

Let's test the function to make sure it can make an `EntitySet` from a data partition.

In [11]:
es1_dict = entityset_from_partition('../input/partitions/p1')
es1_dict['es']

Entityset: clients
  Entities:
    app [Rows: 3458, Columns: 122]
    bureau [Rows: 16097, Columns: 17]
    previous [Rows: 16204, Columns: 37]
    bureau_balance [Rows: 166374, Columns: 4]
    cash [Rows: 96632, Columns: 8]
    installments [Rows: 129130, Columns: 8]
    credit [Rows: 35694, Columns: 23]
  Relationships:
    bureau.SK_ID_CURR -> app.SK_ID_CURR
    bureau_balance.SK_ID_BUREAU -> bureau.SK_ID_BUREAU
    previous.SK_ID_CURR -> app.SK_ID_CURR
    cash.SK_ID_PREV -> previous.SK_ID_PREV
    installments.SK_ID_PREV -> previous.SK_ID_PREV
    credit.SK_ID_PREV -> previous.SK_ID_PREV

The function works as intended. The next step is to write a function that can take a single `EntitySet` and the `features` we want to build, and make a feature matrix. (`entityset_from_partition` returns a dictionary with the partition number so we can save the feature matrix based on this number.)

# Function to Create Feature Matrix from EntitySet 

With the entity set and the feature names, generating the feature matrix is a one-liner in Featuretools. Since we are going to use Dask for parallelizing the operation, we'll set the number of jobs to 1. The `chunk_size` is an extremely important parameter, and I'd suggest experimenting with this to find the optimal value. What I've found works best is setting the `chunk_size` to the length of the entire dataset provided it can all fit in memory. The best choice may depend on the exact problem. 

The last step in the function is to save the feature matrix to disk using the name of the partition of data. Using the `feature_defs` ensures that we create the exact same set of features for each parition.

In [12]:
def feature_matrix_from_entityset(es_dict, feature_defs, return_fm = False):
    """Run deep feature synthesis from an entityset and feature definitions. 
    Saves feature matrix based on partition.""" 
    
    # Extract the entityset
    es = es_dict['es']
    
    # Calculate the feature matrix and save
    feature_matrix = ft.calculate_feature_matrix(feature_defs, 
                                                 entityset=es, 
                                                 n_jobs = 1, 
                                                 verbose = 0,
                                                 chunk_size = es['app'].df.shape[0])
    
    feature_matrix.to_csv('../input/fm/p%d_fm.csv' % es_dict['num'], index = True)
    
    if return_fm:
        return feature_matrix

Below we test the function using the entityset from the first partition.

In [13]:
import warnings
warnings.filterwarnings('ignore', category = FutureWarning)

start = timer()
fm1 = feature_matrix_from_entityset(es1_dict, feature_defs, return_fm = True)
end = timer()
fm1.shape

(3458, 1820)

In [14]:
print(f'Computing one feature matrix took {round(end - start, 2)} seconds.')

Computing one feature matrix took 82.1 seconds.


__We now have both parts needed to go from a data partition on disk to a feature matrix made using 1/104 of the data.__ All we have to do is repeat this operation 104 times and we will have all of our features. Since we have eight cores, we can make eight feature matrices at once (your number may differ). This gets around the fundamental bottleneck in the calculation: __running on a single core is inefficient__ especially when we have 8 available. 

To actually run this in parallel, we use the Dask library.

# Dask

We will use the Dask to parallelize the calculation of feature matrices. First, we'll import and set up a `Client` using processes, which will create one worker for each core on the machine. The memory limit of each worker will be the total system memory (16 gb) divided by the number of cores (8). 

Then we'll use the `db.from_sequence` method to create a "Dask bag" from the partition paths. A [Dask bag](http://dask.pydata.org/en/latest/bag-overview.html) is just a set of operations that we want to run in parallel. We then `map` the paths to the `entityset_from_partition` function which will create the `EntitySets`. These in turn are `map`ped to the `feature_matrix_from_entityset` to make the `feature_matrix` for one of the 104 partitions. (A map is a method that takes a function and a list of inputs and applys the function to each input).

Below we clear the system memory for a full run of Dask.

In [15]:
import gc

# Free up all system memory
gc.enable()
del app, bureau, bureau_balance, previous, credit, cash, installments
gc.collect()

322

The code below starts up 8 workers, each using one of our cores. The memory limit per worker will be the total system memory divided by the number of cores. We use `processes` instead of threads because we doing computationally heavy work and because each calculation is __independent__ meaning the workers do not need to talk with one another. 

(The issue with Python and threads is that threads share memory - processes do not - and because of the [Global Interpreter Lock](https://wiki.python.org/moin/GlobalInterpreterLock) there are few operations that can run in parallel using threads.) For more on the topic of processes, threads, and the global interpreter lock in Python, I recommend [this article](https://medium.com/@bfortuner/python-multithreading-vs-multiprocessing-73072ce5600b).

In [16]:
import dask.bag as db
from dask.distributed import Client

# Use all 8 cores
client = Client(processes = True)

In [17]:
client.ncores()

{'tcp://127.0.0.1:59289': 1,
 'tcp://127.0.0.1:59292': 1,
 'tcp://127.0.0.1:59295': 1,
 'tcp://127.0.0.1:59297': 1,
 'tcp://127.0.0.1:59300': 1,
 'tcp://127.0.0.1:59301': 1,
 'tcp://127.0.0.1:59303': 1,
 'tcp://127.0.0.1:59305': 1}

## Visualizations of Dask

After starting a `Client`, if you have `Bokeh` installed, you can navigate to http://localhost:8787/ to view the status of the workers. Doing this on my machine (8 cores with 16 gb total RAM) gives me:

![](../images/process_workers.png)

Right now we aren't taxing our system very much! 

Next, let's create a list of paths of our partitions.

In [18]:
paths = ['../input/partitions/p%d' %  i for i in range(1, 105)]
paths[:8]

['../input/partitions/p1',
 '../input/partitions/p2',
 '../input/partitions/p3',
 '../input/partitions/p4',
 '../input/partitions/p5',
 '../input/partitions/p6',
 '../input/partitions/p7',
 '../input/partitions/p8']

We made the partitions small enough that none of the feature matrices will be too large for an individual worker. 

The next step is the heart of the code. We create a "Dask bag" from the paths, map this to the `EntitySet` creating function, and then map the result to the `feature_matrix` create and save function. The `EntitySet` is never saved and only exists in working memory. The cell below does not actually execute the code, but only creates the `bag` of tasks that Dask will then be able to allocate to our workers. 

In [19]:
# Create a bag object
b = db.from_sequence(paths)

# Map entityset function
b = b.map(entityset_from_partition)

# Map feature matrix function
b = b.map(feature_matrix_from_entityset, feature_defs = feature_defs)
    
b

dask.bag<map-fea..., npartitions=104>

The cell below carries out the computation. Nothing is returned since each feature matrix is saved as a `csv`.

In [20]:
overall_start = timer()
b.compute()
overall_end = timer()

print(f"Total Time Elapsed: {round(overall_end - overall_start, 2)} seconds.")

distributed.nanny - WARNING - Worker process 66950 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker


Total Time Elapsed: 2371.93 seconds.


If we look at the task graph in the Bokeh dashboard we can see all the tasks that and the sequence in which they have to be completed. From the structure of this [Directed Acyclic Graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph), it's clear that this problem is highly parallizable! In computer science terms, this is referred to as [embarassingly parallel](http://www.cs.iusb.edu/~danav/teach/b424/b424_23_embpar.html) because the tasks do not need to communicate with one another. This also lets us run on processes which do not share memory.

![](../images/taskgraph_inprogress.png)

If you have Bokeh installed, you can see system information during the run. 

For example, we can look at the task stream which shows the workers and all the tasks:

![](../images/taskstream_inprogress.png)

All 8 of our cores are being utilized with a total of 208 tasks to complete.

The profile tab shows the amount of time taken by each operation:

![](../images/profile2.png)

As the operations complete, the Task Graph updates.

![](../images/taskgraph_completing.png)

## Optional Final Feature Matrix

If we want one final matrix, we can read in the individual feature matrices and join them together. This could be done in Dask using threads, but it just as easily can be done in pure Python with Pandas.

In [21]:
# Base directory for feature matrices
base = '../input/fm/'
fm_paths = [base + p for p in os.listdir(base) if 'fm.csv' in p]

First we read in the dataframes and place them in a list.

In [22]:
read_start = timer()
fms = [pd.read_csv(path) for path in fm_paths]
read_end = timer()

print(f'Reading in {len(fms)} feature matrices took {round(read_end - read_start)} seconds.')

Reading in 104 feature matrices took 141 seconds.


Then we concatenate all the dataframes in the list along the first axis - meaning that we add the rows to each other.

In [23]:
concat_start = timer()
feature_matrix = pd.concat(fms, axis = 0)
concat_end = timer()

print('Final Feature Matrix Shape:', feature_matrix.shape)

Final Feature Matrix Shape: (356255, 1821)


In [24]:
print(f"Concatenation time: {round(concat_end - concat_start, 2)} seconds.")

Concatenation time: 13.51 seconds.


The final feature matrix is exactly the expected shape: the number of clients in `app` by the number of features. 

If you don't already have the feature matrix, you can use the following line to save it to disk. This is now ready for feature selection and modeling! 

In [25]:
# feature_matrix.reset_index(inplace = True)
# feature_matrix.to_csv('../input/feature_matrix.csv', index = False)
feature_matrix.head()

SK_ID_CURR  AMT_ANNUITY  AMT_CREDIT  AMT_GOODS_PRICE  AMT_INCOME_TOTAL  \
0      116121       7033.5    126000.0         126000.0           45000.0   
1      116122      26316.0    900000.0         900000.0           81000.0   
2      116123      21996.0    454500.0         454500.0           90000.0   
3      116124      27801.0    543735.0         486000.0           90000.0   
4      116126      29700.0    462694.5         418500.0          216000.0   

   AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_HOUR  \
0                        0.0                         0.0   
1                        NaN                         NaN   
2                        0.0                         0.0   
3                        NaN                         NaN   
4                        0.0                         0.0   

   AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
0                        0.0                        0.0   
1                        NaN                        NaN   
2                        0.0                        1.0   
3                        NaN                        NaN   
4                        0.0                        2.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  \
0                         0.0   
1                         NaN   
2                         0.0   
3                         NaN   
4                         0.0   

                          ...                          \
0                         ...                           
1                         ...                           
2                         ...                           
3                         ...                           
4                         ...                           

   PERCENTILE(MEAN(credit.AMT_TOTAL_RECEIVABLE))  \
0                                            NaN   
1                                       0.180308   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   PERCENTILE(MEAN(credit.CNT_DRAWINGS_ATM_CURRENT))  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

   PERCENTILE(MEAN(credit.CNT_DRAWINGS_CURRENT))  \
0                                            NaN   
1                                       0.173191   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   PERCENTILE(MEAN(credit.CNT_DRAWINGS_OTHER_CURRENT))  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   PERCENTILE(MEAN(credit.CNT_DRAWINGS_POS_CURRENT))  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

   PERCENTILE(MEAN(credit.CNT_INSTALMENT_MATURE_CUM))  \
0                                                NaN    
1                                           0.182681    
2                                                NaN    
3                                                NaN    
4                                                NaN    

   PERCENTILE(MEAN(credit.SK_DPD))  PERCENTILE(MEAN(credit.SK_DPD_DEF))  \
0                              NaN                                  NaN   
1                         0.434164                             0.450178   
2                              NaN                               

# Conclusions

__Using Dask to run the operations in parallel on all our cores reduced the total time to get the feature matrix from 25 hours to under 3 hours.__ This notebook shows that sometimes instead of always trying to get a bigger machine, we need to think about how to use the resouces we have as efficiently as possible. 

For this problem, to run it in parallel, we had to break one large problem into several smaller ones as follows:

1. Partition the data into subsets based on the clients
2. Write a function to generate an `EntitySet` from a partition
3. Write a function to create a `feature_matrix` from an `EntitySet`
4. Set up Dask to use all 8 cores to make a feature matrix from 8 partitions at once
5. Save the resulting feature matrices to disk
6. (Optional) read in the individual feature matrices to create one final feature matrix.

__Parallel processing allows us to take full advantage of our system's resources and the same framework developed in this notebook can be applied to other data science and machine learning problems as well: break a large problem up into manageable smaller ones. Now, not only can we use Featuretools to automatically generate thousands of relevant features, but we can use Dask to run the calculation in parallel and get the most out of our available system.__